Stats: Insights: Add annual site stats

psealock opened the pull request:
Add yearly stats as a stats module to Insights page

<img width="458" alt="screen shot 2017-12-28 at 10 21 51 pm" src="https://user-images.githubusercontent.com/1922453/34406616-f3194668-ec1e-11e7-8ad9-1ccfdb29c1fe.png">

cc @jancavan 

### Notes
* Average Comments is taken to one decimal place. Maybe this should only be for values < 10 or something?
* "View All" link at the bottom of the module can be included in the next PR that creates that page

### ToDo
- [x] Finalize design and update CSS and markup
- [x] i18n strings
- [x] Fix failing tests around the `utils` modification

jancavan commented:
> should we have a fallback state if/when no data has been returned by the API?

Yes. We can go with "No annual stats recorded". Thoughts?

a8ci18n commented on client/my-sites/stats/annual-site-stats/index.js:
```diff
@@ -0,0 +1,113 @@
+					<div className="annual-site-stats__stat-figure is-large">{ data.total_posts }</div>
+				</div>
+				<div className="annual-site-stats__stat">
+					<div className="annual-site-stats__stat-title">{ translate( 'total comments' ) }</div>
```
Hi! I've found a possible matching string that has already been translated **35** times:
`translate( 'Total comments:' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

psealock commented:
### Update
* Tests now pass
* i18n strings
* Updated to match recent API changes
* Added "No annual stats recorded for this year." message

@jancavan FWIW here's the mobile view without collapsing any stats. Maybe its ok as is?

![screen shot 2017-12-29 at 2 15 20 pm](https://user-images.githubusercontent.com/1922453/34427097-0a523c70-eca3-11e7-81c9-a34c1878e9ee.png)

timmyc commented on client/my-sites/stats/annual-site-stats/index.js:
```diff
@@ -0,0 +1,113 @@
+		siteId: PropTypes.number,
+		years: PropTypes.array,
+		translate: PropTypes.func,
+		moment: PropTypes.func,
```
minor but missing `statType` here.

timmyc commented on client/my-sites/stats/annual-site-stats/index.js:
```diff
@@ -0,0 +1,113 @@
+				</div>
+				<div className="annual-site-stats__stat">
+					<div className="annual-site-stats__stat-title">{ translate( 'total posts' ) }</div>
+					<div className="annual-site-stats__stat-figure is-large">{ data.total_posts }</div>
```
All the numeric values should be passed through `this.props.numberFormat()` to localize the thousands separator and such.

timmyc commented on client/my-sites/stats/annual-site-stats/index.js:
```diff
@@ -0,0 +1,113 @@
+		/* eslint-disable wpcalypso/jsx-classname-namespace */
+		return (
+			<div>
+				{ siteId && <QuerySiteStats siteId={ siteId } statType={ statType } /> }
```
If we are only going to use this component in the context of the Insights page, we could reasonably rely upon the "Post Trends" component already making the query to the `statsInsights` endpoint. On this branch that endpoint is called 2x now on the page load with the same query params.

If we consider adding this component to the Years period page, then we would need to add the query component to ensure we have data to work with.

psealock commented:
Thanks for the review @timmyc 
https://github.com/Automattic/wp-calypso/pull/21126/commits/5feb864f31fcc95b3ec264a431adf7de5ae9520e adds `numberFormat`, fixes propTypes, and removes extra query.

>On this branch that endpoint is called 2x now on the page load with the same query params.

Good catch. I've removed the query for now. We can add it back in if needed later.

>When loading the Insights page on an empty cache - the following is shown while the data is being fetched:

This is weird because the query component is rendered but doesn't immediately fire a request. I'm not sure why, but other components on the page suffer the same affliction because initially there is no data but `isRequesting === false`.

a8ci18n commented on client/my-sites/stats/annual-site-stats/index.js:
```diff
@@ -0,0 +1,115 @@
+					</div>
+				</div>
+				<div className="annual-site-stats__stat">
+					<div className="annual-site-stats__stat-title">{ translate( 'total comments' ) }</div>
```
Hi! I've found a possible matching string that has already been translated **35** times:
`translate( 'Total comments:' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

psealock commented:
https://github.com/Automattic/wp-calypso/pull/21126/commits/5f6dfb1cf5f89ad63e83212bad83b863aa811d7e determines loading state by presence of data instead of explicitly checking if the request is occurring. `noData` and `isError` states are also introduced.

@timmyc It should be more seamless while data is being fetched and/or waiting for the browser to fire a request.

a8ci18n commented on client/my-sites/stats/annual-site-stats/index.js:
```diff
@@ -0,0 +1,113 @@
+				<SectionHeader label={ translate( 'Annual Site Stats' ) } />
+				<Card className="stats-module">
+					<StatsModulePlaceholder isLoading={ isLoading } />
+					{ isError && <ErrorPanel message={ translate( 'Oops, something went wrong' ) } /> }
```
Hi! I've found a possible matching string that has already been translated **29** times:
`translate( 'Oops! Something went wrong.' )` *ES Score: 21*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/annual-site-stats)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jancavan commented:
> should we consider moving it to the right column under Followers instead?

I think this module makes more sense to be next to "All time" so I think we should keep its located as is.

I've updated the CSS on mobile in: https://github.com/Automattic/wp-calypso/pull/21126/commits/39e2a9f6de0210ac6bf54b3ea38f408cb673532e

![screenshot 2017-12-29 22 39 45](https://user-images.githubusercontent.com/4924246/34451905-3c034ef8-ece9-11e7-912f-8a47249d31d4.png)

beaulebens commented:
I noticed that now that we're into a new year, I get no results, and no way to view the previous year (which is actually the more interesting data!)

![screen shot 2018-01-02 at 4 44 42 pm](https://user-images.githubusercontent.com/108942/34504558-79518360-efdc-11e7-9ad4-544c519e8431.png)

timmyc commented:
@beaulebens I _think_ the plan is to add in support for a "summary view" type page for this stat module which will allow viewing prior year's stats after this initial PR ships. It will be accessed via a "View All" link at the bottom of the module on the insights page:

![view-all-0011](https://user-images.githubusercontent.com/22080/34506147-03e59280-efdf-11e7-8492-b7966d2d6c00.png)

via p10gg3-8qN-p2

blowery commented:
@beaulebens @timmyc Alternatively, maybe we show the previous year's stats while we're in January of the following year? Or show both for some range of time?

blowery commented:
One possibility:

<img width="353" alt="screenshot 2018-01-03 10 10 16" src="https://user-images.githubusercontent.com/14350/34525857-5a4b5b70-f06e-11e7-9447-79836587df1d.png">

blowery commented:
Added the second year while in january in https://github.com/Automattic/wp-calypso/pull/21126/commits/38235fb9655a832a355de2ea2f7f23bf0026f60b

a8ci18n commented on client/my-sites/stats/annual-site-stats/index.js:
```diff
@@ -0,0 +1,120 @@
+				<SectionHeader label={ translate( 'Annual Site Stats', { args: [ currentYear ] } ) } />
+				<Card className="stats-module">
+					<StatsModulePlaceholder isLoading={ isLoading } />
+					{ isError && <ErrorPanel message={ translate( 'Oops, something went wrong' ) } /> }
```
Hi! I've found a possible matching string that has already been translated **29** times:
`translate( 'Oops! Something went wrong.' )` *ES Score: 21*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/annual-site-stats)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

timmyc commented:
@psealock is out until the 10th so if we are feeling good about this @jancavan I think we should get it out there for stats fans.

blowery commented:
@timmyc Jan is also out until the the 12th. I think we should just ship it.

timmyc commented:
@blowery gotcha I think we can give it a merge too. One item though, with the addition of the second year, the mobile view looks like:

![mobile-2-years](https://user-images.githubusercontent.com/22080/34532478-1f5cf6b6-f06b-11e7-8944-52cc1a7fba3f.png)

🤔  wondering if we should make the year row bold to help delineate the two

blowery commented:
<img width="413" alt="screenshot 2018-01-03 13 26 58" src="https://user-images.githubusercontent.com/14350/34533812-d897836c-f089-11e7-9673-33d482946d04.png">

how's that?

timmyc commented:
Looks good to me :shipit:

psealock commented:
Thanks @jancavan @timmyc @blowery for adopting this one and seeing it through. Looking great
